### ISSUE TITLE
Set transaction isolation level

### ISSUE THREAD

**Issue by tgriesser** [knex/knex#581]
It'd be nice to have a way to specify the isolation level on the transaction

Also mentioned by @llambda in #138, it'd be nice to have both a better api for injecting connection specific config default settings (currently it's possible with an `afterCreate` hook)

**Comment by rprieto** [comment]
This would definitely be nice :+1: 
AFAIK, the current workaround looks something like (for `Postgres`):

```js
knex.transaction(function(trx) {
   return trx
   .raw('set transaction isolation level repeatable read;')
   .then(function() {
      return trx.select(/* ... */);
   });
})
```

Does this look correct to you?
And were you thinking of something like these?

```js
knex.transaction({isolation: 'repeatable read'}, function(trx) { /* ... */ });
knex.transaction.repeatableRead(function(trx) { /* ... */ });
```

**Comment by bendrucker** [comment]
Your workaround looks correct. As for the new feature, I'd go with the first version but invert the arguments. Callback first, options second.

**Comment by rprieto** [comment]
Thanks! I suggested the callback as the last argument for CoffeeScript / ES6 support, where the `(trx) =>` function syntax makes it much easier if there's nothing behind it. But you'd have to add argument type checking to keep it backwards-compatible. I guess otherwise we'll just extract that function and call `knex.transaction(fn, {isolation: '...'})`.

**Comment by cspotcode** [comment]
@rprieto's workaround won't work for MySQL, since "SET TRANSACTION ISOLATION LEVEL" applies to the *next* transaction, not the current one.

From: https://dev.mysql.com/doc/refman/5.0/en/set-transaction.html
> Without any `SESSION` or `GLOBAL` keyword, the statement sets the isolation level for the next (not started) transaction performed within the current session. Subsequent transactions revert to using the `SESSION` isolation level.

**Comment by hendrul** [comment]
:( This is a blocker for me, and is sad cuz I really like bookshelf/knex combination. If there where some work around to this for mysql I'll be very happy.

**Comment by honestserpent** [comment]
Any news on this feature?

**Comment by elhigu** [comment]
@rprieto's suggestion for the backwards compatible API looks good to me.

**Comment by honestserpent** [comment]
Any news on this feature?

**Comment by kibertoad** [comment]
@honestserpent I could try working on it. Which databases are you interested in? Would you consider contributing sample SQLs to illustrate what expected output is?

**Comment by honestserpent** [comment]
I'm working on Postgres.
As far as the sample SQL that I would expect is simply: `begin transaction isolation level repeatable read;`

**Comment by Recodify** [comment]
For mysql you can simply do:

```
 await knex.raw('SET TRANSACTION ISOLATION LEVEL READ COMMITTED;');
        await knex.transaction(async (trx) => {
              ...
```

This will set the transaction isolation level to READ COMMITTED for the next transaction only.
See the section 'Transaction Characteristic' on: https://dev.mysql.com/doc/refman/8.0/en/set-transaction.html

**Comment by elhigu** [comment]
@Recodify That doesn't work unless you manage to make sure that those both queries are sent through same connection to DB server. Knex doesn't guarantee that even if pool size is 1 (it might automatically refresh broken/timed out connection).

**Comment by Recodify** [comment]
@elhigu you're right! If I print the current isolation level I can see that indeed it still prints the default

```
const isolationLevel = await this.knex.raw('SELECT @@tx_isolation as result;').transacting(trx);
            console.log(isolationLevel[0][0].result); // prints REAPTABLE READ
```

The strange thing is that it DOES change the behavior of the following code...which is what led me to believe it worked. I'm missing something, will continue to investigate

**Comment by elhigu** [comment]
With some luck it works, but it is not guaranteed to work. So it may fail for example if there are more concurrent queries executed at the same time. If it works it just means that knex did return the same connection for both of the queries.

**Comment by Recodify** [comment]
It has to be more than luck, I've just hit it with 5000 concurrent requests with connection pool settings of: pool: { min: 0, max: 20 } and it works as expected with the behavior clearly that of READ COMMITTED, removing the line that sets this isolation level results in a pretty fast fail due to REPEATABLE READ behavior.

**Comment by elhigu** [comment]
In next knex version you can setup listeners to pool internal operations and see why it returns always the same connection for both of the queries in that case. You can also try break it more by adding `await sleep(10);` between those two queries to simulate more extreme conditions. You can also check to which connection queries are sent from debug prints when you start knex with environment variable DEBUG=knex:* set.

**Comment by elhigu** [comment]
Anyways this is all really unrelated to this issues, which is about having real support by knex to set the isolation level.

**Comment by juicetin** [comment]
Is there a workaround that would enable a `trx` object to be passed around?
E.g.

```
const a = async (trx) => {
  return knex
    .transacting(trx)
    .etc...
}

const b = async (trx) => {
  return knex
    .transacting(trx)
    .etc...
}

const c = async () => {
  // the part that wouldn't/doesn't actually work
  const trx = knex.transaction(trx => trx.raw('SET TRANSACTION ISOLATION LEVEL READ COMMITTED;'))
  await a(trx);
  await b(trx);
};
```

**Comment by elhigu** [comment]
@juicetin check `knex.transactionProvider()` docs.

**Comment by chapmanjacobd** [comment]
For mssql I think nested transaction _might_ work?? But knex seems to only use one transaction at any time.

```ts
const readIsolationLevelSQL = `
SELECT CASE transaction_isolation_level
WHEN 0 THEN 'Unspecified'
WHEN 1 THEN 'ReadUncommitted'
WHEN 2 THEN 'ReadCommitted'
WHEN 3 THEN 'Repeatable'
WHEN 4 THEN 'Serializable'
WHEN 5 THEN 'Snapshot' END AS TRANSACTION_ISOLATION_LEVEL
FROM sys.dm_exec_sessions
where session_id = @@SPID
`;

(async () => {
await mssql.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;');
  await mssql.transaction(async (trxWrapper) => {
    await mssql.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;');
    await trxWrapper.raw(readIsolationLevelSQL).then(console.log);

    await trxWrapper.raw('select @@TRANCOUNT').then(console.log);

    await trxWrapper.transaction(async (trx) => {
      await mssql.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;');
      await trx.raw('select @@TRANCOUNT').then(console.log);
      await trx.raw(readIsolationLevelSQL).then(console.log);
    });
  });
});
```

everything is `1` and `[ { TRANSACTION_ISOLATION_LEVEL: 'ReadCommitted' } ]`

**Comment by elhigu** [comment]
@chapmanjacobd for that code does it matter that the first query might go to a different connection / session that the query which starts the transaction?

```
  await mssql.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;');
  await mssql.transaction(async (trxWrapper) => {
... 
```

Also that configuration disables parameter bindings, so the app might be vulnerable to SQL injection.

**Comment by chapmanjacobd** [comment]
@elhigu yes it would matter. I was just showing that even if it goes on the same connection (with luck) then it still doesn't work for some reason. I'm not sure why. When I do the same thing in dbeaver it works. I will do some more testing

> that configuration disables parameter bindings

which part specifically ?? `.transaction(async (trxWrapper) => {` ?? or you are just talking about `.raw()` ?

**Comment by elhigu** [comment]
I was looking this one:

> multipleStatements : true

On mysql driver it just renders all parameter bindings to SQL string, which renders all ? bindings vulnerable. https://github.com/mysqljs/mysql#multiple-statement-queries

Anyways I just notices that you are actually using mssql, which doesn't even have that option available. So I  have no idea if multiple statements even work at all on mssql.

**Comment by ayroblu** [comment]
Okay, so how do I do this with mysql, mssql? The `set transaction isolation level` must be done prior to the begin transaction step, and the `knex.transaction()` doesn't support a connection parameter to force it to be on the same connection.

**Comment by ayroblu** [comment]
Implemented as `knex.transaction({isolationLevel: 'repeatable read'})` and similar FYI

**Comment by Venryx** [comment]
For PostgreSQL, I'm currently using this and it works fine:
```
await transaction.raw("SET CONSTRAINTS ALL DEFERRED;"); // call this before the other actions within the transaction
```

Is my code above equivalent to `knex.transaction({isolationLevel: 'repeatable read'})`? If not, what are the differences?

**Comment by ayroblu** [comment]
> For PostgreSQL, I'm currently using this and it works fine:
> 
> ```
> await transaction.raw("SET CONSTRAINTS ALL DEFERRED;"); // call this before the other actions within the transaction
> ```
> 
> Is my code above equivalent to `knex.transaction({isolationLevel: 'repeatable read'})`?
> If not, what are the differences?

Isolations level are very specific ideas relating to what happens when other transactions try to reference the same data that you're using in your current transaction, basically from no checks to full table locks / aborts

The deferred constraint checking is whether checks occur immediately or at the end of the transaction

**Comment by jordaoqualho** [comment]
In case you need to specify an isolation level for your transaction, you can use a config parameter isolationLevel. Not supported by oracle and sqlite, options are read uncommitted, read committed, repeatable read, snapshot (mssql only), serializable.

const isolationLevel = 'read committed';
const trx = await knex.transaction({isolationLevel});
const result1 = await trx(tableName).select();
await knex(tableName).insert({ id: 1, value: 1 });
const result2 = await trx(tableName).select();
await trx.commit();

**Comment by remram44** [comment]
I came here from Google but it seems the feature is implemented by #4185 (`knex.transaction({isolationLevel})`). Perhaps this issue can be closed?

**Comment by julianCast** [comment]
Doing something like:
<img width="684" alt="image" src="https://github.com/knex/knex/assets/559677/27f47646-6e42-4f0a-8b65-b7c97d27e6d4">

Causes this:
```
 TypeError: container is not a function

      at node_modules/knex/lib/execution/transaction.js:233:22
```
What am I doing wrong?

**Comment by remram44** [comment]
Where are you calling knex?

**Comment by julianCast** [comment]
> Where are you calling knex?

Im passing it as a CB from the objection Model.
    `return DBModel.transaction((trx) => transactionSequence(trx));`
whereas trx is a Knex.Transaction

**Comment by remram44** [comment]
You set the isolation level wherever you get that Transaction from knex.

**Comment by mercmobily** [comment]
Good news everyone! This feature was implemented back in December 2020 via PR #4185 and has been available since Knex v0.95.0!

Transaction isolation levels are now fully supported using the `isolationLevel` configuration parameter:

```javascript
// Using async/await
const trx = await knex.transaction({ isolationLevel: 'repeatable read' });

// Using callbacks
await knex.transaction({ isolationLevel: 'serializable' }, async (trx) => {
  // your transaction code here
});
```

**Supported isolation levels:**
- `'read uncommitted'`
- `'read committed'`
- `'repeatable read'`
- `'snapshot'` (MSSQL only)
- `'serializable'`

**Database support:**
-  PostgreSQL
-  MySQL/MariaDB
-  MSSQL
-  Redshift
- NO -  SQLite (not needed - always serialized)
- NO -  Oracle (not supported)

The feature is fully documented in our [Transactions Guide](https://knexjs.org/guide/transactions.html#transaction-modes).

Thanks to everyone who contributed to the discussion over the years, especially @ayroblu for the implementation and @kibertoad for the review. Closing this issue as completed.

For anyone arriving here from search engines: the feature you're looking for is already available! Check the documentation link above for usage examples.
